src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs

```diff
@@ -167,30 +179,40 @@ internal static void RecycleSocketAsyncEventArgs(SocketAsyncEventArgs args)
                 }
             }
         }
-        /// <summary>
-        /// Releases all resources owned by the connection
-        /// </summary>
-        protected virtual void Dispose(bool disposing)
+
+        public async Task DisposeAsync()
         {
-            if (disposing)
+            if (!_disposed)
             {
+                _disposed = true;
                 _stopping = true;
                 _output.Reader.CancelPendingRead();
 
-                Task.WaitAll(_sendTask, _receiveTask);
+                await Task.WhenAll(_sendTask, _receiveTask);
```

Comment:
Doesn't look like `_sendTask` or `_receiveTask` will throw as they catch almost everything, and have an empty catch for the rest; still on't tell you about it though.

`AsyncVoidMethodBuilder` captures the current `SynchronizationContext` and posts stuff back to that, starting run, completion, exception etc; whereas `AsyncTaskMethodBuilder` stuffs in the Task.

Feels like code smell though; but I not know enough about `async void` other than a having a general aversion from use; but it does interact with the sync context in a way `async Task` doesn't seem to which means it may behave differently depending on runtime environment?